### PR TITLE
Add more tests to Metric SDK aggregation

### PR DESCRIPTION
**Pull request by cijothomas**
Adding more tests to test meter and instrument identify part, and whether they get merged or not.

### CODE REVIEW

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 1]
```diff
@@ -214,6 +216,178 @@ mod tests {
...
+        // Arrange
+        let exporter = InMemoryMetricsExporter::default();
+        let reader = PeriodicReader::builder(exporter.clone(), runtime::Tokio).build();
+        let meter_provider = SdkMeterProvider::builder().with_reader(reader).build();
+
+        // Act
+        // Meters are identical except for scope attributes, but scope attributes are not an identifying property.
+        // Hence there should be a single metric stream output for this test.
+        let meter1 = meter_provider.versioned_meter(
+            "test.meter1",
```
Comment (lalitb) [inline opentelemetry-sdk/src/metrics/mod.rs:323]:
```suggestion
            "test.meter",
```

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 2]
```diff
@@ -214,6 +216,178 @@ mod tests {
...
+            Some("v0.1.0"),
+            Some("schema_url"),
+            Some(vec![KeyValue::new("key", "value1")]),
+        );
+        let meter2 = meter_provider.versioned_meter(
+            "test.meter1",
```
Comment (lalitb) [inline opentelemetry-sdk/src/metrics/mod.rs:329]:
```suggestion
            "test.meter",
```

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 3]
```diff
@@ -214,6 +216,178 @@ mod tests {
...
+            resource_metrics[0].scope_metrics.len() == 1,
+            "There should be a single scope as the meters are identical"
+        );
+        assert!(
+            resource_metrics[0].scope_metrics[0].metrics.len() == 1,
+            "There should be single metric for the scope as instruments are identical"
+        );
+
+        let scope = &resource_metrics[0].scope_metrics[0].scope;
+        assert_eq!(scope.name, "test.meter1");
```
Comment (lalitb) [inline opentelemetry-sdk/src/metrics/mod.rs:367]:
```suggestion
        assert_eq!(scope.name, "test.meter");
```

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 4]
```diff
@@ -214,6 +216,178 @@ mod tests {
...
+        assert_eq!(scope.version, Some(Cow::Borrowed("v0.1.0")));
+        assert_eq!(scope.schema_url, Some(Cow::Borrowed("schema_url")));
+
+        // Should we validate this, as this is a user error
+        assert_eq!(scope.attributes, vec![KeyValue::new("key", "value1")]);
```
Comment (lalitb) [inline opentelemetry-sdk/src/metrics/mod.rs:372]: Can let it be here. I think specs doesn't say anything whether the attributes should get overwritten or not/.

**Review by lalitb — APPROVED** [review]
Thanks for the tests.

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 5]
```diff
@@ -214,6 +216,178 @@ mod tests {
         assert_eq!(datapoint.value, 15);
     }
 
+    // "multi_thread" tokio flavor must be used else flush won't
+    // be able to make progress!
```
Comment (TommyCpp) [inline opentelemetry-sdk/src/metrics/mod.rs:220]: nit: don't think we need this for every test but not a strong opinion

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 6]
```diff
@@ -214,6 +216,178 @@ mod tests {
...
+
+        let counter2 = meter2
+            .u64_counter("my_counter")
+            .with_unit(Unit::new("my_unit"))
+            .with_description("my_description")
+            .init();
+
+        let attribute = vec![KeyValue::new("key1", "value1")];
+        counter1.add(10, &attribute);
+        counter2.add(5, &attribute);
```
Comment (TommyCpp) [inline opentelemetry-sdk/src/metrics/mod.rs:245]: We can just loop over `test.meter1` and `test.meter2` here?

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 7]
```diff
@@ -214,6 +216,178 @@ mod tests {
...
+                .expect("Sum aggregation expected for Counter instruments by default");
+
+            // Expecting 1 time-series.
+            assert_eq!(sum2.data_points.len(), 1);
+
+            let datapoint2 = &sum2.data_points[0];
+            assert_eq!(datapoint2.value, 5);
+        } else {
+            panic!("No MetricScope found for 'test.meter2'");
+        }
```
Comment (TommyCpp) [inline opentelemetry-sdk/src/metrics/mod.rs:307]: We can probably simplify it a little bit if we use for loop here.

**Review by TommyCpp — COMMENTED** [review]
the test looks good. I'd add some more details on what we are testing here. 

I personally like table driven unit tests so if we can use for loop here may save us some duplication

File: opentelemetry-sdk/src/metrics/mod.rs (hunk 7 above)
Comment (cijothomas) [reply opentelemetry-sdk/src/metrics/mod.rs:307]: agree. i am intentionally writing these tests without moving things to helper methods as I don't have a full grasp yet on how much we should leverage these tests vs something else. Once I get better idea, will move to helper methods wherever applicable, and remove all the code duplication!

**Pull request merged**